### PR TITLE
Start of the manual

### PR DESCRIPTION
**Pull request by diml**
This PR sets up a new manual for ppxlib and fills in the two first sections:

- What is ppx?
- Ppx for end users

It also simplifies the readme. Some of the old contents was either too complicated, not relevant anymore or should go in the manual.

### CODE REVIEW

File: doc/ppx-for-end-users.rst [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+.. code:: sh
+
+          $ ocamlc -dsource _build/default/src/foo.pp.ml
+
+[@@deriving_inline]
+-------------------
+
+Ppxlib supports attaching the ``[@@deriving]`` attribute to type
+declaration. This is used to generate code at compile time based on
```
Comment (xclerc) [inline doc/ppx-for-end-users.rst]: ("declaration" -> "declarations")

File: doc/ppx-for-end-users.rst [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+``[@@deriving <derivers>]`` by ``[@@deriving_inline
+<derivers>][@@@end]``. Then run the following command:
+
+.. code:: sh
+
+          $ dune build --auto-promote
+
+If you reload the file in your editor, you should now see the contents
+of the generated code between the ``[@@deriving_inline]`` and
+``[@@@end]`` attribute. This can help understanding what is provided
```
Comment (xclerc) [inline doc/ppx-for-end-users.rst]: Typo: "attribute" -> "attributes".

File: doc/ppx-for-end-users.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+by a ppx rewriter or debug compilation errors.
+
+Dropping ppx dependencies with [@@deriving_inline]
+--------------------------------------------------
+
+You might notice that the resulting file when using
+``[@@deriving_inline]`` needs no special treatment to be compiled. In
+particular, you can build it without the ppx rewriter or even
+ppxlib. You only need them while developping the project, in order to
+automatically the generated code but that's it. End users of your
```
Comment (xclerc) [inline doc/ppx-for-end-users.rst]: A word seems to be missing after "automatically".

File: doc/what-is-ppx.rst [hunk 4]
```diff
@@ -0,0 +1,119 @@
+************
+What is PPX?
+************
+
+Overview
+--------
+
+Ppx is a meta-programming system for the OCaml programming
+language. It allows developpers to generate code at compile time in a
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "developpers" -> "developers".

Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "strcutured" -> "structured".

File: doc/what-is-ppx.rst [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+How does it works?
+------------------
+
+The ppx system is composed of 3 parts:
+
+- individual ppx rewriters
+- ppxlib
+- a hook in the compiler
+
+Inidividual ppx rewriters are those implemented by various developpers
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "Inidividual" -> "Individual".

Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "developpers" -> "developers".

File: doc/what-is-ppx.rst [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+to provide features to end users, such as ppx_expect_ which provides a
+good inline testing framework.
+
+All these rewriters are written against the ppxlib API. Ppxlib is
+responsible for acknowledging the various rewriters a end user wants
+to use, making sure they can be composed together and performing the
+actual rewriting of input files.
+
+The hook in the compiler allows ppxlib to insert itself in the
+compilation pipeline and perform the rewriting of input files based on
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: ("peform" -> "performs"?)

File: doc/what-is-ppx.rst [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+between the two is as follow:
+
+- ``-pp`` takes as argument a command that is used to parse the
+  textual representation. Such a command can produce either a plain
+  OCaml source file or a serialised representation of the AST
+
+- ``-ppx`` takes as argument a command that is given a serialised
+  representation of the AST and produces another serialised AST
+
+Ppxlib generally uses the first one as it yields faster compilation
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: I would think it is the second one.

File: doc/what-is-ppx.rst [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+times, however it supports both methods of operation.
+
+Is ppxlib necessary?
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: "necessary" seems to be quite a strong word, given
that the next paragraph states that one can write ppx
rewriters without it.

File: doc/what-is-ppx.rst [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+hooks directly, doing so is strongly discouraged for the following
+reasons:
+
+- composing such ppx rewriters is slow and yields much slower
+  compilation times
+- the ABI of the hook is not stable and regularly changes in
+  incompatible ways. This means that a ppx rewriter using the compiler
+  hook directly is likely to work only with a single version of the
+  OCaml compiler
+- the compiler does not provide a good composition semantic, which
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "semantic" -> "semantics".

File: doc/what-is-ppx.rst [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  means that input files will not always be transformed as
+  expected. It is hard to predict what the final result will be, and
+  for end users it is hard to understand what is happening when things
+  go wrong
+- the compiler doesn't handle hygiene: if an attribute is mystyped or
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "mystyped" -> "mistyped".

File: doc/what-is-ppx.rst [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+In summary, ppxlib abstracts away from all the low-level details of
+the ppx system and exposes a consistent model to authors of ppx
+rewriters and end users.
+
+Current state of the ppx ecosystem
+----------------------------------
+
+Ppxlib was developed after the introduction of the ppx system. As a
+result many ppx rewriters do not currently use ppxlib and are using
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: (Coma after "result"?)

File: doc/what-is-ppx.rst [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+the compiler hooks directly. Ppxlib can acknowledge such rewriter so
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "rewriter" -> "rewriters".

File: doc/what-is-ppx.rst [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+that they can be used in conjunction with more moderm rewriters,
```
Comment (xclerc) [inline doc/what-is-ppx.rst]: Typo: "moderm" -> "modern".

**Comment by xclerc** [comment]
Meta-comment: we should probably use something like
[Grammarly](https://www.grammarly.com/) to proofread
the documentation.

**Comment by NathanReb** [comment]
I spent quite some time working with ppxlib recently but still got the fresh eyes of a newbie. I'd really love to help with that but I'm not sure exactly how I should/could do it.

If this was merged then I guess we could incrementally contribute.

What do you think?

**Comment by diml** [comment]
That sounds good to, help on this would definitely be welcome! I'm going to address @xclerc comments and merge the PR

File: README.md [hunk 14]
```diff
@@ -9,495 +7,12 @@ This readme is still under construction.
...
--------
-
-`Ppx_traverse` can generate the following classes: `map`, `iter`, `fold`,
-`fold_map`, `map_with_context`, `lift`. `[@@deriving traverse]` is an alias
-to generate all the supported classes.
-
-`lift` is a special class that is mostly useful to lift an OCaml constant to
-the AST that represent this constant. To do so, you can use
-`Ppx_metaquot_lifters`:
+# Overview
```
[the remark this answers is not in the document]
Comment (diml) [reply README.md]: Agreed. I'm just going to put in a separate file to keep the README.md shorter.

File: doc/index.rst [hunk 15]
```diff
@@ -0,0 +1,35 @@
+ppxlib's user manual
+====================
+
+Overview
+--------
+
+This is the user manual for ppxlib, the core of the ppx
+metaprogramming system for OCaml_ and its derivatives such as
```
[the remark this answers is not in the document]
Comment (diml) [reply doc/index.rst]: Indeed. I'm using meta-programming

**Pull request merged**